# Data Cleaning and Imputation

The purpose of this notebook is to clean the 2015 BRFSS dataset. On Kaggle, we found the following data cleaning notebook: https://www.kaggle.com/code/alexteboul/heart-disease-health-indicators-dataset-notebook/notebook.

This notebook did a lot of what we wanted to do to the data (e.g. change features to binary, make feature headings more comprehensible, etc), however it removed too much data, as it would remove any row which had even a single piece of missing data. Therefore, we edited the notebook (sections 1-4) to make sure it removed less missing data (which will be imputed later).

Then in section 5, we split the data into a test and training set. In section 6, we impute the sets.

## 1. Get the data

In [1]:
#importing necessary modules
import os
import pandas as pd
import random
from numpy.random import RandomState
rng = RandomState()
random.seed(1)

In [2]:
#importing full dataset
brfss_2015_dataset = pd.read_csv('2015.csv')
#Download link available here: https://drive.google.com/file/d/1yUZmDZttzFwHRpSu0v9BjZr3a-mm1f_8/view

In [3]:
#How many rows and columns
brfss_2015_dataset.shape

(441456, 330)

In [4]:
#check that the data loaded in is in the correct format
pd.set_option('display.max_columns', 500)
brfss_2015_dataset.head()

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE         SEQNO  \
0     1.0     1.0  b'01292015'  b'01'  b'29'  b'2015'    1200.0  2.015000e+09   
1     1.0     1.0  b'01202015'  b'01'  b'20'  b'2015'    1100.0  2.015000e+09   
2     1.0     1.0  b'02012015'  b'02'  b'01'  b'2015'    1200.0  2.015000e+09   
3     1.0     1.0  b'01142015'  b'01'  b'14'  b'2015'    1100.0  2.015000e+09   
4     1.0     1.0  b'01142015'  b'01'  b'14'  b'2015'    1100.0  2.015000e+09   

           _PSU  CTELENUM  PVTRESD1  COLGHOUS  STATERES  CELLFON3  LADULT  \
0  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
1  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
2  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
3  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
4  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   

   NUMADULT        NUMMEN  NUMWOMEN  CTELNUM1  CELLFON2  CADULT  PVTRESD2  \
0       3.0  1.000000e+00       2.0       NaN       NaN     NaN       NaN   
1       1.0  5.397605e-79       1.0       NaN       NaN     NaN       NaN   
2       2.0  1.000000e+00       1.0       NaN       NaN     NaN       NaN   
3       3.0  1.000000e+00       2.0       NaN       NaN     NaN       NaN   
4       2.0  1.000000e+00       1.0       NaN       NaN     NaN       NaN   

   CCLGHOUS  CSTATE  LANDLINE  HHADULT  GENHLTH  PHYSHLTH  MENTHLTH  POORHLTH  \
0       NaN     NaN       NaN      NaN      5.0      15.0      18.0      10.0   
1       NaN     NaN       NaN      NaN      3.0      88.0      88.0       NaN   
2       NaN     NaN       NaN      NaN      4.0      15.0      88.0      88.0   
3       NaN     NaN       NaN      NaN      5.0      30.0      30.0      30.0   
4       NaN     NaN       NaN      NaN      5.0      20.0      88.0      30.0   

   HLTHPLN1  PERSDOC2  MEDCOST  CHECKUP1  BPHIGH4  BPMEDS  BLOODCHO  CHOLCHK  \
0       1.0       1.0      2.0       1.0      1.0     1.0       1.0      1.0   
1       2.0       1.0      1.0       4.0      3.0     NaN       1.0      4.0   
2       1.0       2.0      2.0       1.0      3.0     NaN       1.0      1.0   
3       1.0       2.0      1.0       1.0      1.0     1.0       1.0      1.0   
4       1.0       1.0      2.0       1.0      3.0     NaN       1.0      1.0   

   TOLDHI2  CVDINFR4  CVDCRHD4  CVDSTRK3  ASTHMA3  ASTHNOW  CHCSCNCR  \
0      1.0       2.0       2.0       2.0      1.0      1.0       2.0   
1      2.0       2.0       2.0       2.0      2.0      NaN       2.0   
2      1.0       7.0       2.0       1.0      2.0      NaN       2.0   
3      1.0       2.0       2.0       2.0      2.0      NaN       2.0   
4      2.0       2.0       2.0       2.0      2.0      NaN       2.0   

   CHCOCNCR  CHCCOPD1  HAVARTH3  ADDEPEV2  CHCKIDNY  DIABETE3  DIABAGE2  SEX  \
0       2.0       1.0       1.0       1.0       2.0       3.0       NaN  2.0   
1       2.0       2.0       2.0       2.0       2.0       3.0       NaN  2.0   
2       1.0       2.0       1.0       2.0       2.0       3.0       NaN  2.0   
3       1.0       2.0       1.0       1.0       2.0       3.0       NaN  2.0   
4       2.0       2.0       1.0       2.0       2.0       3.0       NaN  2.0   

   MARITAL  EDUCA  RENTHOM1  NUMHHOL2  NUMPHON2  CPDEMO1  VETERAN3  EMPLOY1  \
0      1.0    4.0       1.0       2.0       NaN      1.0       2.0      8.0   
1      2.0    6.0       1.0       2.0       NaN      2.0       2.0      3.0   
2      2.0    4.0       1.0       2.0       NaN      1.0       2.0      7.0   
3      1.0    4.0       1.0       2.0       NaN      1.0       2.0      8.0   
4      1.0    5.0       1.0       2.0       NaN      2.0       2.0      8.0   

   CHILDREN  INCOME2  INTERNET  WEIGHT2  HEIGHT3  PREGNANT  QLACTLM2  \
0      88.0      3.0       2.0    280.0    510.0       NaN       1.0   
1      88.0      1.0       1.0    165.0    508.0       NaN       1.0   
2      88.0     9

**At this point we have 441,456 records and 330 columns. Each record contains an individual's BRFSS survey responses.**

In [5]:
# select specific columns
brfss_df_selected = brfss_2015_dataset[['_MICHD', 
                                         '_RFHYPE5',  
                                         'TOLDHI2', '_CHOLCHK', 
                                         '_BMI5', 
                                         'SMOKE100', 
                                         'CVDSTRK3', 'DIABETE3', 
                                         '_TOTINDA', 
                                         '_FRTLT1', '_VEGLT1', 
                                         '_RFDRHV5', 
                                         'HLTHPLN1', 'MEDCOST', 
                                         'GENHLTH', 'MENTHLTH', 'PHYSHLTH', 'DIFFWALK', 
                                         'SEX', '_AGEG5YR', 'EDUCA', 'INCOME2' ]]

In [6]:
brfss_df_selected.shape

(441456, 22)

In [7]:
brfss_df_selected.head()

_MICHD  _RFHYPE5  TOLDHI2  _CHOLCHK   _BMI5  SMOKE100  CVDSTRK3  DIABETE3  \
0     2.0       2.0      1.0       1.0  4018.0       1.0       2.0       3.0   
1     2.0       1.0      2.0       2.0  2509.0       1.0       2.0       3.0   
2     NaN       1.0      1.0       1.0  2204.0       NaN       1.0       3.0   
3     2.0       2.0      1.0       1.0  2819.0       2.0       2.0       3.0   
4     2.0       1.0      2.0       1.0  2437.0       2.0       2.0       3.0   

   _TOTINDA  _FRTLT1  _VEGLT1  _RFDRHV5  HLTHPLN1  MEDCOST  GENHLTH  MENTHLTH  \
0       2.0      2.0      1.0       1.0       1.0      2.0      5.0      18.0   
1       1.0      2.0      2.0       1.0       2.0      1.0      3.0      88.0   
2       9.0      9.0      9.0       9.0       1.0      2.0      4.0      88.0   
3       2.0      1.0      2.0       1.0       1.0      1.0      5.0      30.0   
4       2.0      9.0      1.0       1.0       1.0      2.0      5.0      88.0   

   PHYSHLTH  DIFFWALK  SEX  _AGEG5YR  EDUCA  INCOME2  
0      15.0       1.0  2.0       9.0    4.0      3.0  
1      88.0       2.0  2.0       7.0    6.0      1.0  
2      15.0       NaN  2.0      11.0    4.0     99.0  
3      30.0       1.0  2.0       9.0    4.0      8.0  
4      20.0       2.0  2.0       9.0    5.0     77.0

## 2. Clean the data

### 2.2 Modify and clean the values to be more suitable to ML algorithms
In order to do this part, I referenced the [codebook](https://www.cdc.gov/brfss/annual_data/2015/pdf/codebook15_llcp.pdf) which says what each column/feature/question is: 

We ensure any binary features are either 0 or 1.

In [8]:
#Replace all missing values with a -1 placeholder
brfss_df_selected.fillna(-1, inplace = True)
brfss_df_selected.head()

C:\Users\shah_\anaconda3\lib\site-packages\pandas\core\frame.py:4317: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


_MICHD  _RFHYPE5  TOLDHI2  _CHOLCHK   _BMI5  SMOKE100  CVDSTRK3  DIABETE3  \
0     2.0       2.0      1.0       1.0  4018.0       1.0       2.0       3.0   
1     2.0       1.0      2.0       2.0  2509.0       1.0       2.0       3.0   
2    -1.0       1.0      1.0       1.0  2204.0      -1.0       1.0       3.0   
3     2.0       2.0      1.0       1.0  2819.0       2.0       2.0       3.0   
4     2.0       1.0      2.0       1.0  2437.0       2.0       2.0       3.0   

   _TOTINDA  _FRTLT1  _VEGLT1  _RFDRHV5  HLTHPLN1  MEDCOST  GENHLTH  MENTHLTH  \
0       2.0      2.0      1.0       1.0       1.0      2.0      5.0      18.0   
1       1.0      2.0      2.0       1.0       2.0      1.0      3.0      88.0   
2       9.0      9.0      9.0       9.0       1.0      2.0      4.0      88.0   
3       2.0      1.0      2.0       1.0       1.0      1.0      5.0      30.0   
4       2.0      9.0      1.0       1.0       1.0      2.0      5.0      88.0   

   PHYSHLTH  DIFFWALK  SEX  _AGEG5YR  EDUCA  INCOME2  
0      15.0       1.0  2.0       9.0    4.0      3.0  
1      88.0       2.0  2.0       7.0    6.0      1.0  
2      15.0      -1.0  2.0      11.0    4.0     99.0  
3      30.0       1.0  2.0       9.0    4.0      8.0  
4      20.0       2.0  2.0       9.0    5.0     77.0

In [9]:
# _MICHD
#Change 2 to 0 because this means did not have MI or CHD
#Removes any data without information on heart disease
brfss_df_selected['_MICHD'] = brfss_df_selected['_MICHD'].replace({2: 0})
brfss_df_selected = brfss_df_selected[brfss_df_selected._MICHD != -1]
brfss_df_selected._MICHD.unique()

<ipython-input-9-c5d5e7fdfd39>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  brfss_df_selected['_MICHD'] = brfss_df_selected['_MICHD'].replace({2: 0})


array([0., 1.])

In [10]:
#1 _RFHYPE5
#Change 1 to 0 so it represents No high blood pressure and 2 to 1 so it represents high blood pressure
brfss_df_selected['_RFHYPE5'] = brfss_df_selected['_RFHYPE5'].replace({1:0, 2:1})
brfss_df_selected['_RFHYPE5'] = brfss_df_selected['_RFHYPE5'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._RFHYPE5 != 9]
brfss_df_selected._RFHYPE5.unique()

array([ 1.,  0., -1.])

In [11]:
#2 TOLDHI2
# Change 2 to 0 because it is No
brfss_df_selected['TOLDHI2'] = brfss_df_selected['TOLDHI2'].replace({2:0})
brfss_df_selected['TOLDHI2'] = brfss_df_selected['TOLDHI2'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.TOLDHI2 != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.TOLDHI2 != 9]
brfss_df_selected.TOLDHI2.unique()

array([ 1.,  0., -1.])

In [12]:
#3 _CHOLCHK
# Change 3 to 0 and 2 to 0 for Not checked cholesterol in past 5 years
brfss_df_selected['_CHOLCHK'] = brfss_df_selected['_CHOLCHK'].replace({3:0,2:0})
brfss_df_selected['_CHOLCHK'] = brfss_df_selected['_CHOLCHK'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._CHOLCHK != 9]
brfss_df_selected._CHOLCHK.unique()

array([ 1.,  0., -1.])

In [13]:
#4 _BMI5 (no changes, just note that these are BMI * 100. So for example a BMI of 4018 is really 40.18)
brfss_df_selected['_BMI5'] = brfss_df_selected['_BMI5'].div(100).round(0)
brfss_df_selected._BMI5.unique()

array([ 40.,  25.,  28.,  24.,  27.,  15.,  30.,  26.,  -0.,  23.,  34.,
        31.,  51.,  14.,  33.,  21.,  22.,  35.,  38.,  20.,  19.,  32.,
        46.,  41.,  37.,  36.,  29.,  39.,  18.,  54.,  45.,  47.,  43.,
        55.,  49.,  42.,  17.,  16.,  48.,  44.,  50.,  59.,  52.,  53.,
        57.,  63.,  65.,  58.,  61.,  13.,  56.,  60.,  74.,  62.,  64.,
        66.,  73.,  68.,  67.,  85.,  72.,  71.,  81.,  84.,  70.,  82.,
        79.,  92.,  75.,  69.,  12.,  87.,  77.,  88.,  96.,  95.,  78.,
        91.,  76.,  89.,  98.,  83.,  86.,  80.,  90., 100.,  97.])

In [14]:
#5 SMOKE100
# Change 2 to 0 because it is No
brfss_df_selected['SMOKE100'] = brfss_df_selected['SMOKE100'].replace({2:0})
brfss_df_selected['SMOKE100'] = brfss_df_selected['SMOKE100'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.SMOKE100 != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.SMOKE100 != 9]
brfss_df_selected.SMOKE100.unique()

array([ 1.,  0., -1.])

In [15]:
#6 CVDSTRK3
# Change 2 to 0 because it is No
brfss_df_selected['CVDSTRK3'] = brfss_df_selected['CVDSTRK3'].replace({2:0})
brfss_df_selected['CVDSTRK3'] = brfss_df_selected['CVDSTRK3'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.CVDSTRK3 != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.CVDSTRK3 != 9]
brfss_df_selected.CVDSTRK3.unique()

array([ 0.,  1., -1.])

In [16]:
#7 DIABETE3
# going to make this ordinal. 0 is for no diabetes or only during pregnancy, 1 is for pre-diabetes or borderline diabetes, 2 is for yes diabetes
brfss_df_selected['DIABETE3'] = brfss_df_selected['DIABETE3'].replace({2:0, 3:0, 1:2, 4:1})
brfss_df_selected['DIABETE3'] = brfss_df_selected['DIABETE3'].replace({2:0, 3:0, 1:2, 4:1})
brfss_df_selected['DIABETE3'] = brfss_df_selected['DIABETE3'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.DIABETE3 != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.DIABETE3 != 9]
brfss_df_selected.DIABETE3.unique()

array([ 0.,  2., -1.])

In [17]:
#8 _TOTINDA
# 1 for physical activity
# change 2 to 0 for no physical activity
brfss_df_selected['_TOTINDA'] = brfss_df_selected['_TOTINDA'].replace({2:0})
brfss_df_selected['_TOTINDA'] = brfss_df_selected['_TOTINDA'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._TOTINDA != 9]
brfss_df_selected._TOTINDA.unique()

array([ 0.,  1., -1.])

In [18]:
#9 _FRTLT1
# Change 2 to 0. this means no fruit consumed per day. 1 will mean consumed 1 or more pieces of fruit per day 
brfss_df_selected['_FRTLT1'] = brfss_df_selected['_FRTLT1'].replace({2:0})
brfss_df_selected['_FRTLT1'] = brfss_df_selected['_FRTLT1'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._FRTLT1 != 9]
brfss_df_selected._FRTLT1.unique()

array([ 0.,  1., -1.])

In [19]:
#10 _VEGLT1
# Change 2 to 0. this means no vegetables consumed per day. 1 will mean consumed 1 or more pieces of vegetable per day 
brfss_df_selected['_VEGLT1'] = brfss_df_selected['_VEGLT1'].replace({2:0})
brfss_df_selected['_VEGLT1'] = brfss_df_selected['_VEGLT1'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._VEGLT1 != 9]
brfss_df_selected._VEGLT1.unique()

array([ 1.,  0., -1.])

In [20]:
#11 _RFDRHV5
# Change 1 to 0 (1 was no for heavy drinking). change all 2 to 1 (2 was yes for heavy drinking)
brfss_df_selected['_RFDRHV5'] = brfss_df_selected['_RFDRHV5'].replace({1:0, 2:1})
brfss_df_selected['_RFDRHV5'] = brfss_df_selected['_RFDRHV5'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._RFDRHV5 != 9]
brfss_df_selected._RFDRHV5.unique()

array([ 0., -1.,  1.])

In [21]:
#12 HLTHPLN1
# 1 is yes, change 2 to 0 because it is No health care access
brfss_df_selected['HLTHPLN1'] = brfss_df_selected['HLTHPLN1'].replace({2:0})
brfss_df_selected['HLTHPLN1'] = brfss_df_selected['HLTHPLN1'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.HLTHPLN1 != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.HLTHPLN1 != 9]
brfss_df_selected.HLTHPLN1.unique()

array([ 1.,  0., -1.])

In [22]:
#13 MEDCOST
# Change 2 to 0 for no, 1 is already yes
brfss_df_selected['MEDCOST'] = brfss_df_selected['MEDCOST'].replace({2:0})
brfss_df_selected['MEDCOST'] = brfss_df_selected['MEDCOST'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.MEDCOST != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.MEDCOST != 9]
brfss_df_selected.MEDCOST.unique()

array([ 0.,  1., -1.])

In [23]:
#14 GENHLTH
# This is an ordinal variable that I want to keep (1 is Excellent -> 5 is Poor)
brfss_df_selected['GENHLTH'] = brfss_df_selected['GENHLTH'].replace({7:-1, 9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.GENHLTH != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.GENHLTH != 9]
brfss_df_selected.GENHLTH.unique()

array([ 5.,  3.,  2.,  4.,  1., -1.])

In [24]:
#15 MENTHLTH
# already in days so keep that, scale will be 0-30
# change 88 to 0 because it means none (no bad mental health days)
brfss_df_selected['MENTHLTH'] = brfss_df_selected['MENTHLTH'].replace({88:0})
brfss_df_selected['MENTHLTH'] = brfss_df_selected['MENTHLTH'].replace({77:-1,99:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.MENTHLTH != 77]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.MENTHLTH != 99]
brfss_df_selected.MENTHLTH.unique()

array([18.,  0., 30.,  3., -1.,  2.,  5., 15., 10.,  1.,  6.,  7., 20.,
       25., 29.,  4.,  8., 21., 14., 26.,  9., 16., 28., 11., 17., 12.,
       23., 24., 13., 27., 22., 19.])

In [25]:
#16 PHYSHLTH
# already in days so keep that, scale will be 0-30
# change 88 to 0 because it means none (no bad mental health days) [physical health days?]
brfss_df_selected['PHYSHLTH'] = brfss_df_selected['PHYSHLTH'].replace({88:0})
brfss_df_selected['PHYSHLTH'] = brfss_df_selected['PHYSHLTH'].replace({77:-1,99:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.PHYSHLTH != 77]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.PHYSHLTH != 99]
brfss_df_selected.PHYSHLTH.unique()

array([15.,  0., 30., 20., -1.,  2., 14.,  5.,  6., 28.,  1.,  7., 10.,
        3., 17.,  4., 19., 21., 25.,  8., 12., 27., 22., 29., 24.,  9.,
       16., 13., 18., 23., 11., 26.])

In [26]:
#17 DIFFWALK
# change 2 to 0 for no. 1 is already yes
brfss_df_selected['DIFFWALK'] = brfss_df_selected['DIFFWALK'].replace({2:0})
brfss_df_selected['DIFFWALK'] = brfss_df_selected['DIFFWALK'].replace({7:-1,9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.DIFFWALK != 7]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.DIFFWALK != 9]
brfss_df_selected.DIFFWALK.unique()

array([ 1.,  0., -1.])

In [27]:
#18 SEX
# in other words - is respondent male (somewhat arbitrarily chose this change because men are at higher risk for heart disease)
# change 2 to 0 (female as 0). Male is 1
brfss_df_selected['SEX'] = brfss_df_selected['SEX'].replace({2:0})
brfss_df_selected.SEX.unique()

array([0., 1.])

In [28]:
#19 _AGEG5YR
# already ordinal. 1 is 18-24 all the way up to 13 wis 80 and older. 5 year increments.
brfss_df_selected['_AGEG5YR'] = brfss_df_selected['_AGEG5YR'].replace({14:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected._AGEG5YR != 14]
brfss_df_selected._AGEG5YR.unique()

array([ 9.,  7., 11., 13., 10., 12.,  8.,  4.,  6.,  2.,  3.,  5.,  1.,
       -1.])

In [29]:
#20 EDUCA
# This is already an ordinal variable with 1 being never attended school or kindergarten only up to 6 being college 4 years or more
# Scale here is 1-6
brfss_df_selected['EDUCA'] = brfss_df_selected['EDUCA'].replace({9:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.EDUCA != 9]
brfss_df_selected.EDUCA.unique()

array([ 4.,  6.,  5.,  3.,  2., -1.,  1.])

In [30]:
#21 INCOME2
# Variable is already ordinal with 1 being less than $10,000 all the way up to 8 being $75,000 or more
brfss_df_selected['INCOME2'] = brfss_df_selected['INCOME2'].replace({77:-1,99:-1})
#brfss_df_selected = brfss_df_selected[brfss_df_selected.INCOME2 != 77]
#brfss_df_selected = brfss_df_selected[brfss_df_selected.INCOME2 != 99]
brfss_df_selected.INCOME2.unique()

array([ 3.,  1.,  8., -1.,  6.,  4.,  7.,  5.,  2.])

In [31]:
#Let's see what the data looks like after Modifying Values
brfss_df_selected.head()

_MICHD  _RFHYPE5  TOLDHI2  _CHOLCHK  _BMI5  SMOKE100  CVDSTRK3  DIABETE3  \
0     0.0       1.0      1.0       1.0   40.0       1.0       0.0       0.0   
1     0.0       0.0      0.0       0.0   25.0       1.0       0.0       0.0   
3     0.0       1.0      1.0       1.0   28.0       0.0       0.0       0.0   
4     0.0       0.0      0.0       1.0   24.0       0.0       0.0       0.0   
5     0.0       1.0      0.0       1.0   27.0       0.0       0.0       0.0   

   _TOTINDA  _FRTLT1  _VEGLT1  _RFDRHV5  HLTHPLN1  MEDCOST  GENHLTH  MENTHLTH  \
0       0.0      0.0      1.0       0.0       1.0      0.0      5.0      18.0   
1       1.0      0.0      0.0       0.0       0.0      1.0      3.0       0.0   
3       0.0      1.0      0.0       0.0       1.0      1.0      5.0      30.0   
4       0.0     -1.0      1.0       0.0       1.0      0.0      5.0       0.0   
5       1.0      1.0      1.0       0.0       1.0      0.0      2.0       0.0   

   PHYSHLTH  DIFFWALK  SEX  _AGEG5YR  EDUCA  INCOME2  
0      15.0       1.0  0.0       9.0    4.0      3.0  
1       0.0       0.0  0.0       7.0    6.0      1.0  
3      30.0       1.0  0.0       9.0    4.0      8.0  
4      20.0       0.0  0.0       9.0    5.0     -1.0  
5       0.0       0.0  0.0      11.0    3.0      6.0

In [32]:
#Check Class Sizes of the heart disease column
brfss_df_selected.groupby(['_MICHD']).size()

_MICHD
0.0    398881
1.0     38633
dtype: int64

## 3. Make feature names more readable

In [33]:
#Rename the columns to make them more readable
brfss = brfss_df_selected.rename(columns = {'_MICHD':'HeartDiseaseorAttack', 
                                         '_RFHYPE5':'HighBP',  
                                         'TOLDHI2':'HighChol', '_CHOLCHK':'CholCheck', 
                                         '_BMI5':'BMI', 
                                         'SMOKE100':'Smoker', 
                                         'CVDSTRK3':'Stroke', 'DIABETE3':'Diabetes', 
                                         '_TOTINDA':'PhysActivity', 
                                         '_FRTLT1':'Fruits', '_VEGLT1':"Veggies", 
                                         '_RFDRHV5':'HvyAlcoholConsump', 
                                         'HLTHPLN1':'AnyHealthcare', 'MEDCOST':'NoDocbcCost', 
                                         'GENHLTH':'GenHlth', 'MENTHLTH':'MentHlth', 'PHYSHLTH':'PhysHlth', 'DIFFWALK':'DiffWalk', 
                                         'SEX':'Sex', '_AGEG5YR':'Age', 'EDUCA':'Education', 'INCOME2':'Income' })

In [34]:
brfss.head()

HeartDiseaseorAttack  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                   0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1                   0.0     0.0       0.0        0.0  25.0     1.0     0.0   
3                   0.0     1.0       1.0        1.0  28.0     0.0     0.0   
4                   0.0     0.0       0.0        1.0  24.0     0.0     0.0   
5                   0.0     1.0       0.0        1.0  27.0     0.0     0.0   

   Diabetes  PhysActivity  Fruits  Veggies  HvyAlcoholConsump  AnyHealthcare  \
0       0.0           0.0     0.0      1.0                0.0            1.0   
1       0.0           1.0     0.0      0.0                0.0            0.0   
3       0.0           0.0     1.0      0.0                0.0            1.0   
4       0.0           0.0    -1.0      1.0                0.0            1.0   
5       0.0           1.0     1.0      1.0                0.0            1.0   

   NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  Education  \
0          0.0      5.0      18.0      15.0       1.0  0.0   9.0        4.0   
1          1.0      3.0       0.0       0.0       0.0  0.0   7.0        6.0   
3          1.0      5.0      30.0      30.0       1.0  0.0   9.0        4.0   
4          0.0      5.0       0.0      20.0       0.0  0.0   9.0        5.0   
5          0.0      2.0       0.0       0.0       0.0  0.0  11.0        3.0   

   Income  
0     3.0  
1     1.0  
3     8.0  
4    -1.0  
5     6.0

In [35]:
brfss.shape

(437514, 22)

In [36]:
#Check how many respondents have had heart disease or a heart attack. Note the class imbalance!
brfss.groupby(['HeartDiseaseorAttack']).size()

HeartDiseaseorAttack
0.0    398881
1.0     38633
dtype: int64

In [37]:
# Turn all -1 placeholders into null entries for later imputation
brfss = brfss.replace(-1, pd.NA)

## 4. Save to csv
We save a .csv where heart disease is the target variable and in the first column.

In [38]:
#************************************************************************************************
brfss.to_csv('heart_disease_90_10.csv', sep=",", index=False)
#************************************************************************************************

## 5. Test and Training Set Split

We decided to split the test and training set into 10% test, 90% training, where each set contains constant proportions of positive and negative cases.

In [39]:
positive = brfss[brfss['HeartDiseaseorAttack'] == 1]
negative = brfss[brfss['HeartDiseaseorAttack'] != 1]

numpos = round(len(positive) * 0.1)
numneg = round(len(negative) * 0.1)

testpos = positive.sample(n=numpos, random_state=rng)
trainpos = positive.loc[~positive.index.isin(testpos.index)]
testneg = negative.sample(n=numneg, random_state=rng)
trainneg = negative.loc[~negative.index.isin(testneg.index)]

testframe = [testpos,testneg]
trainframe = [trainpos,trainneg]

test = pd.concat(testframe)
train = pd.concat(trainframe)

test.shape

(43751, 22)

In [40]:
trainpos.shape

(34770, 22)

In [41]:
trainneg.shape

(358993, 22)

In [42]:
train.shape

(393763, 22)

In [43]:
# Checking 10% of the data is in the test set
a = test.shape[0]
b = train.shape[0]
100 * a/(a+b)

9.999908574354192

We now export these sets.

In [44]:
#************************************************************************************************
test.to_csv('heart_disease_test_90_10.csv', sep=",", index=False)
train.to_csv('heart_disease_train_90_10.csv', sep=",", index=False)
#************************************************************************************************

## 6. Imputation

We decided that we wanted to impute our data, and, since during lectures we had covered how k nearest neighbours techniques could be used for imputation, we decided to implement a KNN imputation. For this we followed [A Guide To KNN Imputation](https://medium.com/@kyawsawhtoon/a-guide-to-knn-imputation-95e2dc496e). As we did not want to focus on imputation for this project, we did not look into other methods or how, within KNN imputation, we could improve our process. For example, alternative distance metrics could be investigated. To save time, we used a package which runs the imputation without much instruction; essentially the only thing that had to be done was to normalise the data, which is needed in order to compare distances.

We now impute the train and test datasets.

In [46]:
from sklearn.impute import KNNImputer
from sklearn.preprocessing import MinMaxScaler

In [47]:
scaler = MinMaxScaler()
imputer = KNNImputer(n_neighbors=5)

In [ ]:
#Test Imputation
test_normalised_df = pd.DataFrame(scaler.fit_transform(test), columns = test.columns)
test_imputed_df = pd.DataFrame(imputer.fit_transform(test_normalised_df),columns = test_normalised_df.columns)
test_imputed_df.to_csv('imputed_test_90_10.csv', sep=",", index=False)

We then uploaded this file to our google drive and github (because it is small enough unlike the training set). 

In [48]:
#Train Imputation
train_normalised_df = pd.DataFrame(scaler.fit_transform(train), columns = train.columns)
train_imputed_df = pd.DataFrame(imputer.fit_transform(train_normalised_df),columns = train_normalised_df.columns)
train_imputed_df.to_csv('imputed_train_90_10.csv', sep=",", index=False)

We then uploaded this file to our google drive.

With this, we have cleaned and imputed the data, ready to use to train and test the models against.